File: Design-Patterns-Using-Kotlin/Adapter/AdapterPatternRectangle.kt
// Desired Shape Interface
interface Shape {
    fun draw()
}

// Circle class
class Circle(private val x: Int, private val y: Int, private val radius: Int) : Shape {
    override fun draw() {
        println("Draw the Circle.")
    }
}

// Rectangle class (Adaptee)
open class Rectangle(private val x: Int, private val y: Int, val length: Int, private val width: Int) {
    fun oldDraw() {
        println("Drawing Rectangle.")
    }
}

// RectangleAdapter class
class RectangleAdapter(x: Int, y: Int, l: Int, w: Int) : Shape {
    private val adaptee = Rectangle(x, y, l, w)

    override fun draw() {
        adaptee.oldDraw()
    }
}

// Client Code
fun main() {
    val adapter: Shape = RectangleAdapter(1, 2, 3, 4)
    adapter.draw()
}

/*
Drawing Rectangle.
*/

File: Design-Patterns-Using-JavaScript/Adapter/AdapterPatternRectangle.js
// Desired Interface
class Shape {
    draw() {
        throw new Error("Abstract method: draw");
    }
}

// Circle class
class Circle extends Shape {
    constructor(x, y, radius) {
        super();
        this.x = x;
        this.y = y;
        this.radius = radius;
    }

    draw() {
        console.log("Draw the Circle.");
    }
}

// Rectangle class (Adaptee)
class Rectangle {
    constructor(x, y, length, width) {
        this.x = x;
        this.y = y;
        this.length = length;
        this.width = width;
    }

    oldDraw() {
        console.log("Drawing Rectangle.");
    }
}

// RectangleAdapter class
class RectangleAdapter extends Shape {
    constructor(x, y, length, width) {
        super();
        this.adaptee = new Rectangle(x, y, length, width);
    }

    draw() {
        this.adaptee.oldDraw();
    }
}

// Client Code
const adapter = new RectangleAdapter(1, 2, 3, 4);
adapter.draw();

/*
Drawing Rectangle.
*/